libsolidity/analysis/ConstantEvaluator.cpp

```diff
@@ -41,22 +41,30 @@ void ConstantEvaluator::endVisit(BinaryOperation const& _operation)
 	auto right = type(_operation.rightExpression());
 	if (left && right)
 	{
-		auto commonType = left->binaryOperatorResult(_operation.getOperator(), right);
-		if (!commonType)
-			m_errorReporter.fatalTypeError(
-				_operation.location(),
-				"Operator " +
-				string(Token::toString(_operation.getOperator())) +
-				" not compatible with types " +
-				left->toString() +
-				" and " +
-				right->toString()
-			);
+		TypeResult result = left->binaryOperatorResult(_operation.getOperator(), right);
+		if (!result)
+		{
+			if (result.error().empty())
+				m_errorReporter.fatalTypeError(
+					_operation.location(),
+					"Operator " +
+					string(Token::toString(_operation.getOperator())) +
+					" not compatible with types " +
+					left->toString() +
+					" and " +
+					right->toString()
+				);
+			else
+				m_errorReporter.fatalTypeError(
+					_operation.location(),
+					result.error()
```

Comment:
Does it make sense to replace the error message completely?

Reply:
It could also make sense to just append the optional error message, which in turn would make the resulting error pretty long. But, the type checker already generates such errors: ``Type rational_const 1 / 5221...(1225 digits omitted)...5168 is not implicitly convertible to expected type int256. Try converting to type ufixed8x80 or use an explicit conversion.``. So we could also do that here.

Reply:
Yes, I would propose to start like this. At some later point we could also structure the error messages (and then output them on separate lines):

 - error
 - reason
 - change suggestion

Reply:
Ok, sounds good!